drain3/template_miner.py

```diff
@@ -173,21 +173,28 @@ def match(self, log_message: str, full_search_strategy="never") -> LogCluster:
     def get_parameter_list(self, log_template: str, content: str):
         escaped_prefix = re.escape(self.config.mask_prefix)
         escaped_suffix = re.escape(self.config.mask_suffix)
-        template_regex = re.sub(escaped_prefix + r".+?" + escaped_suffix, self.drain.param_str, log_template)
-        if self.drain.param_str not in template_regex:
-            return []
-        template_regex = re.escape(template_regex)
-        template_regex = re.sub(r'\\ +', r'\\s+', template_regex)
-        template_regex = "^" + template_regex.replace(escaped_prefix + r"\*" + escaped_suffix, "(.*?)") + "$"
+        content_masked =  re.sub(escaped_prefix + r".+?" + escaped_suffix, lambda match: self.config.mask_prefix + "*" + self.config.mask_suffix, content)
+        for instruction in self.config.masking_instructions:
+            content_masked = instruction.regex.sub(lambda match: self.config.mask_prefix + "(" + re.escape(match.group(0)) + ")" + self.config.mask_suffix, content_masked)
+        masking_parameter_list = [re.sub(r"\\(.)", r"\1", match.group(1)) for match in re.finditer(escaped_prefix + r"\((.+?)(?<!\\)\)" + escaped_suffix, content_masked)]
+
+        escaped_param_str = re.escape(self.drain.param_str)
+        template_regex = re.escape(log_template)
+        if escaped_param_str not in template_regex:
+            return masking_parameter_list
+        masking_parameter_iterator = iter(masking_parameter_list)
+        template_regex = re.sub(re.escape(escaped_prefix) + r"(?!\\[*]).+?" + re.escape(escaped_suffix), lambda x: "(" + next(masking_parameter_iterator) + ")", template_regex)
```

Comment:
Only replace occurrences of masks introduced by `MaskingInstruction` with the parameters in order.
(Note: This is where the faulty assumption is made that all masks introduced by `MaskingInstruction` do not have `*` as the value of `mask_with`.)